### PR TITLE
Optimize surface drawing logic

### PR DESCRIPTION
**Pull request by ZehMatt**
There are a lot of optimization opportunities when it comes to branching, using specialization with template arguments allows the compiler to remove anything not necessary for the specific path.

### CODE REVIEW

File: src/openrct2/paint/tile_element/Paint.Surface.cpp [hunk 1]
```diff
@@ -452,7 +453,7 @@ static void viewport_surface_smoothen_edge(
     uint8_t neighbourCorners[2] = { 0 };
     uint8_t ownCorners[2] = { 0 };
 
-    switch (edge)
+    switch (TEdge)
```
Comment (Broxzier) [inline src/openrct2/paint/tile_element/Paint.Surface.cpp:456]: Compilers are probably smart enough to see that only one here is used and inline that, but to guarantee that no branching takes place here, you could use `if constexpr` instead. Unfortunately, `switch constexpr` is not a thing (yet).

File: src/openrct2/paint/tile_element/Paint.Surface.cpp [hunk 2]
```diff
@@ -491,7 +492,7 @@ static void viewport_surface_smoothen_edge(
     }
 
     uint8_t dh = 0, cl = 0;
-    switch (edge)
+    switch (TEdge)
```
Comment (Broxzier) [inline src/openrct2/paint/tile_element/Paint.Surface.cpp:495]: See above.

File: src/openrct2/paint/tile_element/Paint.Surface.cpp [hunk 3]
```diff
@@ -574,7 +575,7 @@ static void viewport_surface_draw_tile_side_bottom(
     CoordsXY tunnelTopBoundBoxOffset = { 0, 0 };
 
     const tunnel_entry* tunnelArray;
-    switch (edge)
+    switch (TEdge)
```
Comment (Broxzier) [inline src/openrct2/paint/tile_element/Paint.Surface.cpp:578]: See above.

File: src/openrct2/paint/tile_element/Paint.Surface.cpp [hunk 4]
```diff
@@ -792,7 +798,7 @@ static void viewport_surface_draw_tile_side_top(
     CoordsXY offset = { 0, 0 };
     CoordsXY bounds = { 0, 0 };
 
-    switch (edge)
+    switch (TEdge)
```
Comment (Broxzier) [inline src/openrct2/paint/tile_element/Paint.Surface.cpp:801]: See above

**Review by Broxzier — COMMENTED** [review]
The changes LGTM. I wouldn't be surprised if compilers were already inlining this and removing branches where possible. Did you measure any performance gain?

**Comment by ZehMatt** [comment]
develop:
```
>openrct2 benchgfx parks/bpb.sv6
Engine: Software
Render Count: 60
Zoom[0] average: 0.089371s, 11 FPS
Zoom[1] average: 0.048420s, 21 FPS
Zoom[2] average: 0.028481s, 35 FPS
Total average: 0.055424s, 18 FPS
Time: 3.32543s
```
pr:
```
>openrct2 benchgfx parks/bpb.sv6
Engine: Software
Render Count: 60
Zoom[0] average: 0.089082s, 11 FPS
Zoom[1] average: 0.047886s, 21 FPS
Zoom[2] average: 0.028248s, 35 FPS
Total average: 0.055072s, 18 FPS
Time: 3.30433s
```
Its a tiny bit better nothing huge.

**Comment by duncanspumpkin** [comment]
I also ran benchgfx on this pr:
develop:
```
Engine: Software
Render Count: 60
Zoom[0] average: 0.167411s, 6 FPS
Zoom[1] average: 0.097102s, 10 FPS
Zoom[2] average: 0.051839s, 19 FPS
Total average: 0.105451s, 9 FPS
Time: 6.32705s
```
this pr:
```
Engine: Software
Render Count: 60
Zoom[0] average: 0.170795s, 6 FPS
Zoom[1] average: 0.096691s, 10 FPS
Zoom[2] average: 0.051904s, 19 FPS
Total average: 0.106463s, 9 FPS
Time: 6.38781s
```
I get +-0.1s on multiple runs so I'd say its within the margin for error i.e. no change.

**Pull request closed without merging**